design/one-pager-helm-provider.md

```diff
@@ -0,0 +1,400 @@
+Kind, we will be able to manage helm releases in **Crossplane provisioned external clusters**, **existing external
+clusters** and also **Crossplane control cluster** (a.k.a. local cluster).
+
+### `Release` Custom Resource and Controller
+
+```
+apiVersion: helm.crossplane.io/v1alpha1
+kind: Release
+metadata:
+  name: wordpress-example
+spec:
+  forProvider:
+    chart:
+      name: wordpress
+      repository: https://charts.bitnami.com/bitnami
+      version: 9.3.19
+    namespace: wordpress
+    values:
+      mariadb:
+        enabled: false
+      externaldb:
+        enabled: true
+    set:
+    - name: wordpressBlogName
+      value: "Hello Crossplane"
+    - name: externalDatabase.host
+      valueFrom:
+        secretKeyRef:
+          name: dbconn
+          key: host
+    - name: externalDatabase.user
+      valueFrom:
+        secretKeyRef:
+          name: dbconn
+          key: username
+    - name: externalDatabase.password
+      valueFrom:
+        secretKeyRef:
+          name: dbconn
+          key: password
+  providerRef: 
+    name: cluster-1-provider
+  reclaimPolicy: Delete
+```
+
+We will implement helm controller using the [managed reconciler of crossplane runtime](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed).
+
+Following logic will be implemented for corresponding functions:
+
+[Connect](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalConnectorFn.Connect):
+
+- Using provided Kubernetes Provider, create a kubernetes client and helm client as `ExternalClient`. 
+
+[Observe](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalClientFns.Observe):
+
+1. Using [history action](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/action/history.go#L43)
+of helm client, get last [release](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/release/release.go#L22)
+information.
+    1. If there is no last release, return [`ExternalObservation`](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalObservation)
+    as `ResourceExists = false` which will result in `Create` to be called.
+    2. If there is last release, compare last [release](https://github.com/helm/helm/blob/3d64c6bb5495d4e4426c27b181300fff45f95ff0/pkg/release/release.go#L22)
+    information with the desired spec (e.g. `spec.forProvider`)
+        1. If desired state matches with observed, return `ResourceUpToDate = true`
+        2. If desired state differs from observed, return `ResourceUpToDate = false`
+
+[Create](https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed#ExternalClientFns.Create):
+
+1. Fetch the helm chart by constructing URL as  `<spec.forProvider.chart.repository>/<spec.forProvider.chart.name>-<spec.forProvider.chart.version>.tgz`
```

Comment:
Adding link to helm repo url: https://chartmuseum.com/docs/#helm-chart-repository

Reply:
Realized that this is not always true and sometimes you may need to append `charts/` after repo URL. Fortunately, helm already has a pull action as well which we can use here instead of dealing with these details. Updated the design doc accordingly.

Good catch, thanks!